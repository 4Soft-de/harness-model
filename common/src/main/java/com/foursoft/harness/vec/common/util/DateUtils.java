/*-
 * ========================LICENSE_START=================================
 * vec-common
 * %%
 * Copyright (C) 2020 - 2022 4Soft GmbH
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * =========================LICENSE_END==================================
 */
package com.foursoft.harness.vec.common.util;

import com.foursoft.harness.vec.common.exception.VecException;

import javax.xml.datatype.DatatypeConfigurationException;
import javax.xml.datatype.DatatypeFactory;
import javax.xml.datatype.XMLGregorianCalendar;
import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZonedDateTime;

/**
 * Utility class for date related operations.
 * <p>
 * Amongst others, it provides methods to create {@link XMLGregorianCalendar}s which is used
 * for date fields within the VEC.
 */
public final class DateUtils {

    private DateUtils() {
        // hide default constructor
    }

    /**
     * Creates a {@link XMLGregorianCalendar} for the current date.
     *
     * @return A never-null {@link XMLGregorianCalendar} for the current date.
     * @throws VecException In case the conversion fails.
     */
    public static XMLGregorianCalendar currentDate() {
        final Instant date = Instant.now();
        return toXMLGregorianCalendar(date.toString());
    }

    /**
     * Converts the passed {@link LocalDateTime} to an {@link XMLGregorianCalendar}.
     *
     * @param date The LocalDate which should be converted.
     * @return A never-null {@link XMLGregorianCalendar} for the given date.
     * @throws VecException In case the conversion fails.
     */
    public static XMLGregorianCalendar toXMLGregorianCalendar(final LocalDate date) {
        return toXMLGregorianCalendar(date.toString());
    }

    /**
     * Converts the passed {@link LocalDateTime} to an {@link XMLGregorianCalendar}.
     *
     * @param dateTime The LocalDateTime which should be converted.
     * @return A never-null {@link XMLGregorianCalendar} for the given date.
     * @throws VecException In case the conversion fails.
     */
    public static XMLGregorianCalendar toXMLGregorianCalendar(final LocalDateTime dateTime) {
        return toXMLGregorianCalendar(dateTime.toString());
    }

    /**
     * Converts the passed date as string to an {@link XMLGregorianCalendar}.
     *
     * @param dateTime The date as string in form {@code yyyyMMdd HH:mm:ss.SSSSSS Z}.
     * @return A never-null {@link XMLGregorianCalendar} for the given date.
     * @throws IllegalArgumentException In case the given input is null or an empty String.
     * @throws VecException             In case the conversion fails.
     */
    public static XMLGregorianCalendar toXMLGregorianCalendar(final String dateTime) {
        if (StringUtils.isEmpty(dateTime)) {
            throw new IllegalArgumentException("Given String may not be empty or null.");
        }

        try {
            return DatatypeFactory.newInstance().newXMLGregorianCalendar(dateTime);
        } catch (final DatatypeConfigurationException | IllegalArgumentException e) {
            throw new VecException(String.format("Failed to create date calender for datetime %s.", dateTime), e);
        }
    }

    /**
     * Converts the passed {@link XMLGregorianCalendar} to a {@link LocalDate}.
     *
     * @param calendar The calendar to convert.
     * @return A never-null {@link LocalDate} for the given calendar.
     */
    public static LocalDate toLocalDate(final XMLGregorianCalendar calendar) {
        return toZonedDateTime(calendar).toLocalDate();
    }

    /**
     * Converts the passed {@link XMLGregorianCalendar} to a {@link LocalDateTime}.
     *
     * @param calendar The calendar to convert.
     * @return A never-null {@link LocalDateTime} for the given calendar.
     */
    public static LocalDateTime toLocalDateTime(final XMLGregorianCalendar calendar) {
        return toZonedDateTime(calendar).toLocalDateTime();
    }

    private static ZonedDateTime toZonedDateTime(final XMLGregorianCalendar calendar) {
        if (calendar == null) {
            throw new IllegalArgumentException("Given Calender may not be null.");
        }

        ZonedDateTime zonedDateTime = calendar.toGregorianCalendar().toZonedDateTime();

        // The nanos are not respected while converting to the GregorianCalendar.
        final BigDecimal fractionalSecond = calendar.getFractionalSecond();
        if (fractionalSecond != null) {
            final String fractionAsString = fractionalSecond.toString();
            // Highest possible value: 0.999999999 -> Can always be converted to a String.
            final int nanos = Integer.parseInt(fractionAsString.substring(2));  // cut of the "0."
            zonedDateTime = zonedDateTime.withNano(nanos);
        }

        return zonedDateTime;
    }

}
