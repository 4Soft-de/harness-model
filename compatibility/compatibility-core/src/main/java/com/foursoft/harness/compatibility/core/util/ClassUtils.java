/*-
 * ========================LICENSE_START=================================
 * compatibility-core
 * %%
 * Copyright (C) 2020 - 2023 4Soft GmbH
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * =========================LICENSE_END==================================
 */
package com.foursoft.harness.compatibility.core.util;

/**
 * Utility class for class determination operations.
 */
public class ClassUtils {

    /**
     * The Object name suffix for generated ByteBuddy-Proxy classes.
     */
    public static final String BYTE_BUDDY = "$ByteBuddy";

    private ClassUtils() {
        // hide
    }

    /**
     * Gets the real class from the given class in case the given class is a ByteBuddy Proxy class.
     *
     * @param clazz Either a "real" class or a ByteBuddy Proxy class.
     * @return The real class.
     */
    public static Class<?> getNonProxyClass(final Class<?> clazz) {
        try {
            return getMappedClass(clazz, clazz.getPackageName());
        } catch (final ClassNotFoundException e) {
            return clazz;  // Should not happen.
        }
    }

    /**
     * Gets the given class but with the given package name.
     * Also returns the real class if a ByteBuddy Proxy class was given.
     *
     * @param clazz       Either a "real" class or a ByteBuddy Proxy class.
     * @param packageName The package name to which the class will be mapped.
     * @return The mapped Class.
     * @throws ClassNotFoundException For non-existing classes.
     */
    public static Class<?> getMappedClass(final Class<?> clazz, final String packageName)
            throws ClassNotFoundException {
        final String simpleNameSuperClass = clazz.getSuperclass().getSimpleName();
        final String simpleName = clazz.getSimpleName();

        final String targetClassName = !simpleNameSuperClass.equals(simpleName) && simpleName.contains(BYTE_BUDDY)
                ? simpleNameSuperClass
                : simpleName;
        final String classWithPackage = String.join(".", packageName, targetClassName);

        return Class.forName(classWithPackage);
    }

}
