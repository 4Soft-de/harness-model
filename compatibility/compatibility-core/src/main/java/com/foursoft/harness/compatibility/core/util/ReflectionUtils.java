/*-
 * ========================LICENSE_START=================================
 * Compatibility Core
 * %%
 * Copyright (C) 2020 - 2023 4Soft GmbH
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * =========================LICENSE_END==================================
 */
package com.foursoft.harness.compatibility.core.util;

import com.foursoft.harness.compatibility.core.exception.WrapperException;

import java.lang.reflect.Field;

/**
 * Utility class for reflection operations.
 */
public final class ReflectionUtils {

    private ReflectionUtils() {
        // hide constructor
    }

    public static void setParentRelationship(final Object object, final String fieldName, final Object parent) {
        try {
            final Field field = org.reflections.ReflectionUtils
                    .getAllFields(object.getClass(),
                                  f -> f.getName().equals(fieldName))
                    .stream()
                    .findFirst().orElse(null);
            if (field == null) {
                final String errorMsg = String.format("Could not find field %s for class %s!",
                                                      fieldName, object.getClass());
                throw new WrapperException(errorMsg);
            }

            setFieldValue(object, field, parent);
        } catch (final IllegalAccessException e) {
            final String errorMsg = String.format("Could not set parent relation (field %s) between %s and %s!",
                                                  fieldName, object.getClass(), parent.getClass());
            throw new WrapperException(errorMsg, e);
        }
    }

    /**
     * Sets the field value for the field of the given target to the given value.
     *
     * @param target Target object which should be accessed / modified.
     * @param field  Field to set.
     * @param value  Value to set for the given field.
     * @throws IllegalAccessException See {@link Field#set(Object, Object)}.
     */
    public static synchronized void setFieldValue(final Object target, final Field field, final Object value)
            throws IllegalAccessException {
        // Ensure the field accessibility is always set to false in the end.
        try {
            field.setAccessible(true);
            field.set(target, value);
        } finally {
            field.setAccessible(false);
        }
    }

}
