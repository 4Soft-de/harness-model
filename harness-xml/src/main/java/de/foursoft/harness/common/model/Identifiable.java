package de.foursoft.harness.common.model;

public interface Identifiable {

    String getXmlId();

}
