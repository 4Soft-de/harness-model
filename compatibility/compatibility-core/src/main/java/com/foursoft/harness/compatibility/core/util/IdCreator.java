/*-
 * ========================LICENSE_START=================================
 * compatibility-core
 * %%
 * Copyright (C) 2020 - 2023 4Soft GmbH
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * =========================LICENSE_END==================================
 */
package com.foursoft.harness.compatibility.core.util;

import java.util.UUID;

/**
 * Utility class which create ids.
 */
public final class IdCreator {

    private IdCreator() {
        // hide constructor
    }

    /**
     * Generate an id for the given object.
     * Will use its class for id creation.
     *
     * @param object Object to get an id for.
     * @return Never-null String containing an id for this object.
     */
    public static String generateXmlId(final Object object) {
        return generateXmlId(object.getClass());
    }

    /**
     * Generate an id for the given class.
     *
     * @param clazz Class to get an id for.
     * @return Never-null String containing an id for this Class.
     */
    public static String generateXmlId(final Class<?> clazz) {
        return String.join("_", "new", clazz.getSimpleName(), UUID.randomUUID().toString());
    }

}
