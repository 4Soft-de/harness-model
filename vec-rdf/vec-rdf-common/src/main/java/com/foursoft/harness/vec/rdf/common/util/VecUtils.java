/*-
 * ========================LICENSE_START=================================
 * VEC RDF Common
 * %%
 * Copyright (C) 2024 4Soft GmbH
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * =========================LICENSE_END==================================
 */
package com.foursoft.harness.vec.rdf.common.util;

import com.foursoft.harness.navext.runtime.model.Identifiable;
import com.foursoft.harness.vec.rdf.common.exception.VecRdfException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.w3c.dom.Document;
import org.xml.sax.SAXException;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.InvocationTargetException;
import java.util.function.Consumer;
import java.util.function.Function;

public class VecUtils {

    private static final Logger LOGGER = LoggerFactory.getLogger(VecUtils.class);

    public static Document loadDocument(InputStream file) {
        try {
            LOGGER.info("Parsing DOM Tree.");
            DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
            documentBuilderFactory.setNamespaceAware(true);
            DocumentBuilder builder = documentBuilderFactory.newDocumentBuilder();
            return builder.parse(file);
        } catch (IOException | ParserConfigurationException | SAXException e) {
            throw new VecRdfException(
                    "Unable to parse XML file with DOM Parser, maybe it is not a well-formed XML file.", e);
        }
    }

    public static void genericVecModelTraversal(Object root, Consumer<Identifiable> consumer) {
        Class<?> rootClass = root.getClass();

        try {
            Class<?> depthFirstTraverserClass = loadVisitorClasses(rootClass, "DepthFirstTraverserImpl");
            Class<?> functionVisitorClass = loadVisitorClasses(rootClass, "FunctionVisitor");
            Class<?> traversingVisitorClass = loadVisitorClasses(rootClass, "TraversingVisitor");
            Class<?> visitorClass = loadVisitorClasses(rootClass, "Visitor");
            Class<?> traverserClass = loadVisitorClasses(rootClass, "Traverser");

            Object depthFirstTraverser = depthFirstTraverserClass.getConstructor()
                    .newInstance();
            Object functionVisitor = functionVisitorClass.getConstructor(Function.class)
                    .newInstance((Function<Identifiable, Void>) (identifiable) -> {
                        consumer.accept(identifiable);
                        return null;
                    });
            Object traverser = traversingVisitorClass.getConstructor(traverserClass, visitorClass)
                    .newInstance(depthFirstTraverser, functionVisitor);

            rootClass.getMethod("accept", visitorClass)
                    .invoke(root, traverser);
        } catch (ClassNotFoundException | InvocationTargetException | InstantiationException | IllegalAccessException |
                 NoSuchMethodException e) {
            throw new VecRdfException("Unable to do generic traversal of VEC DOM due to reflection issues.", e);
        }
    }

    private static Class<?> loadVisitorClasses(Class<?> rootClass, String className) throws ClassNotFoundException {
        return Class.forName(rootClass.getPackageName() + ".visitor." + className, true, rootClass.getClassLoader());
    }

}
