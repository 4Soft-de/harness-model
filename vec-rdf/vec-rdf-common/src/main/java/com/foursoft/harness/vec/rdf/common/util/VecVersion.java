/*-
 * ========================LICENSE_START=================================
 * VEC RDF Common
 * %%
 * Copyright (C) 2024 4Soft GmbH
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * =========================LICENSE_END==================================
 */
package com.foursoft.harness.vec.rdf.common.util;

import com.foursoft.harness.navext.runtime.io.read.XMLReader;
import com.foursoft.harness.navext.runtime.io.write.XMLWriter;
import com.foursoft.harness.navext.runtime.model.Identifiable;
import com.foursoft.harness.vec.rdf.common.exception.VecRdfException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.w3c.dom.Document;
import org.w3c.dom.NodeList;

import java.util.Arrays;
import java.util.Objects;

public enum VecVersion {

    V113("1.1.3"), V120("1.2.0"), V121("1.2.1"), V122("1.2.2"), V200("2.0.0"), V201("2.0.1"), V202("2.0.2"), V210(
            "2.1.0");

    private static final Logger LOGGER = LoggerFactory.getLogger(VecVersion.class);

    private final String version;

    VecVersion(String version) {
        this.version = version;
    }

    public String getVersion() {
        return version;
    }

    public XMLReader<? extends Identifiable, Identifiable> resolveReader() {
        if ("1.1.3".equals(version)) {
            return new com.foursoft.harness.vec.v113.VecReader();
        }
        if (version.startsWith("1.2.")) {
            return new com.foursoft.harness.vec.v12x.VecReader();
        }
        if (version.startsWith("2.")) {
            return new com.foursoft.harness.vec.v2x.VecReader();
        }
        throw new VecRdfException("No VEC Api for Version: '" + version + "' found.");
    }

    public XMLWriter<? extends Identifiable> resolveWriter() {
        if ("1.1.3".equals(version)) {
            return new com.foursoft.harness.vec.v113.VecWriter();
        }
        if (version.startsWith("1.2.")) {
            return new com.foursoft.harness.vec.v12x.VecWriter();
        }
        if (version.startsWith("2.")) {
            return new com.foursoft.harness.vec.v2x.VecWriter();
        }
        throw new VecRdfException("No VEC Api for Version: '" + version + "' found.");
    }

    public static VecVersion guessVersion(Document document) {
        LOGGER.debug("Trying to guess VEC version.");
        NodeList versions = document.getElementsByTagName("VecVersion");
        if (versions.getLength() == 0) {
            throw new VecRdfException("Unable to find 'VecVersion' tag in.");
        }
        String version = versions.item(0)
                .getTextContent();
        LOGGER.info("VEC version detected: {}", version);
        return VecVersion.fromVersion(version);
    }

    public static VecVersion fromVersion(String version) {
        Objects.requireNonNull(version, "version must not be null");
        VecVersion[] enumConstants = VecVersion.class.getEnumConstants();
        return Arrays.stream(enumConstants)
                .filter(v -> version.equals(v.getVersion()))
                .findFirst()
                .orElseThrow(() -> new VecRdfException("Unsupported VEC Version: " + version));
    }
}
