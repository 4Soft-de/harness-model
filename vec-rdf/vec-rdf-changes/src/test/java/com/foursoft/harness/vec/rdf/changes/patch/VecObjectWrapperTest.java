/*-
 * ========================LICENSE_START=================================
 * VEC RDF Changesets (Experimental)
 * %%
 * Copyright (C) 2024 4Soft GmbH
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * =========================LICENSE_END==================================
 */
package com.foursoft.harness.vec.rdf.changes.patch;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.foursoft.harness.navext.runtime.model.Identifiable;
import com.foursoft.harness.vec.v2x.VecNumericalValue;
import com.foursoft.harness.vec.v2x.VecSIUnit;
import com.foursoft.harness.vec.v2x.VecWireLength;
import org.apache.jena.rdf.model.Model;
import org.apache.jena.rdf.model.Resource;
import org.apache.jena.riot.Lang;
import org.apache.jena.riot.RDFDataMgr;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.util.HashMap;

import static com.foursoft.harness.vec.rdf.changes.test.TestUtils.*;
import static org.assertj.core.api.Assumptions.assumeThat;

@ExtendWith({SnapshotExtension.class})
class VecObjectWrapperTest {

    private Expect expect;

    private static final String BNODE1_TTL = """
            [
                rdf:type vec:WireLength ;
                rdfs:label "WireLength[id_00074]" ;
                vec:wireLengthLengthType vec:WireLengthType_DMU ;
                vec:wireLengthLengthValue
                    [
                        rdf:type vec:NumericalValue ;
                        rdfs:label "NumericalValue[id_00075]" ;
                        vec:numericalValueValueComponent
                            "1063.0"^^xsd:double ;
                        vec:valueWithUnitUnitComponent :SIUnit-id_00102 ;
                        vec-dbg:id "id_00075"
                    ] ;
                vec-dbg:id "id_00074"
            ].
            """;

    @Test
    void should_load_values() {
        Model m1 = loadModel(BNODE1_TTL);
        RDFDataMgr.read(m1, loadResourceFromClasspath("/vec/v2.1.0/vec-2.1.0-ontology.ttl"), Lang.TTL);
        Resource bNode1 = findNodeWithXmlId(m1, "id_00074");

        assumeThat(bNode1).isNotNull();

        VecWireLength element = new VecWireLength();
        VecSIUnit unit = new VecSIUnit();
        unit.setXmlId("id_00102 ");

        HashMap<String, Identifiable> uriToIdentifiable = new HashMap<>();
        uriToIdentifiable.put("https://www.foursoft.com/test#SIUnit-id_00102", unit);
        MergeContext context = new MergeContext(element, uriToIdentifiable);

        VecObjectWrapper loader = new VecObjectWrapper(context, element);

        bNode1.listProperties()
                .forEach(s -> loader.loadValue(s.getPredicate(), s.getObject()));

        expect.serializer("json")
                .toMatchSnapshot(element);
    }

    @Test
    void should_unload_values() {
        Model m1 = loadModel(BNODE1_TTL);
        RDFDataMgr.read(m1, loadResourceFromClasspath("/vec/v2.1.0/vec-2.1.0-ontology.ttl"), Lang.TTL);
        Resource bNode1 = findNodeWithXmlId(m1, "id_00074");

        assumeThat(bNode1).isNotNull();

        VecWireLength element = new VecWireLength();
        element.setXmlId("id_00074");
        element.setLengthType("DMU");

        VecSIUnit unit = new VecSIUnit();
        unit.setXmlId("id_00102 ");

        VecNumericalValue numericalValue = new VecNumericalValue();
        numericalValue.setXmlId("id_00075");
        numericalValue.setValueComponent(1063.0);
        numericalValue.setUnitComponent(unit);

        element.setLengthValue(numericalValue);

        HashMap<String, Identifiable> uriToIdentifiable = new HashMap<>();
        uriToIdentifiable.put("https://www.foursoft.com/test#SIUnit-id_00102", unit);
        MergeContext context = new MergeContext(element, uriToIdentifiable);

        VecObjectWrapper loader = new VecObjectWrapper(context, element);

        bNode1.listProperties()
                .forEach(s -> loader.removeValue(s.getPredicate(), s.getObject()));

        expect.serializer("json")
                .toMatchSnapshot(element);
    }

}
