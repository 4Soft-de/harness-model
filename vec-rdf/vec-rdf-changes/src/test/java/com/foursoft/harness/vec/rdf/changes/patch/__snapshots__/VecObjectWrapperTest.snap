com.foursoft.harness.vec.rdf.changes.patch.VecObjectWrapperTest.should_load_values=[
  {
    "refExternalMapping": [ ],
    "lengthType": "DMU",
    "lengthValue": {
      "unitComponent": {
        "xmlId": "id_00102 ",
        "refBuildingBlockSpecification2D": [ ],
        "refBuildingBlockSpecification3D": [ ],
        "refCavityPositionDetail": [ ],
        "refCompositeUnit": [ ],
        "refDimension": [ ],
        "refLocalGeometrySpecification": [ ],
        "refValueWithUnit": [ ]
      },
      "valueComponent": 1063.0
    }
  }
]


com.foursoft.harness.vec.rdf.changes.patch.VecObjectWrapperTest.should_unload_values=[
  {
    "xmlId": "id_00074",
    "refExternalMapping": [ ]
  }
]